### DISCUSSION TITLE
Make Django (single) application portable

### DISCUSSION THREAD

**Discussion by zvolsky** [python-poetry/poetry#3582] [Q&A]
I would like know more about usage of poetry with Django.

Django as whole project has its dependencies typically in a virtual environment so I think it is usual to use poetry for the whole project.

However poetry has build/publish features.
And we usually want make a single django (custom) application portable.
**Is it possible build/publish single application with poetry, while poetry root (.toml+.lock) is at the main project level?**

I think such something is NOT possible:
In the main .toml file to create subsections for applications we want to build separately, and start the poetry build command with optional parameter (application ie. subsection name).

So I made this attempt:
Application inside a project as it is usual. New .toml inside the application directory (project's subdirectory).
But I have had no success with this.

2nd attempt:
Later I have copied the application folder into an empty folder with same name. And inside the outer folder I can make a .toml file and I am able to build. I have no idea if the result is correct.

So probably I could move the application outside from project and install it via poetry in editable mode:
--- A directory (../myapp-with-toml/myapp/) ---
In myapp-with-toml I should then create pyproject.toml designed for the application. And build from there.

Is this a good way or there is a better approach how to build portable Django application using Poetry?

(initially I asked via issues: https://github.com/python-poetry/poetry/issues/3578)

**Comment by sinoroc** [comment]
Can you show a minimal example of what you are trying to achieve: project directory structure (use [_`tree`_](https://linux.die.net/man/1/tree) maybe), skeleton of a `pyproject.toml`, and so on?

**Comment by zvolsky** [comment]
Here is the standard basic Django project structure with poetry initialized.
```
.
├── app1
│   ├── admin.py
│   ├── apps.py
│   ├── __init__.py
│   ├── migrations
│   │   └── __init__.py
│   ├── models.py
│   ├── tests.py
│   └── views.py
├── app2
│   ├── admin.py
│   ├── apps.py
│   ├── __init__.py
│   ├── migrations
│   │   └── __init__.py
│   ├── models.py
│   ├── tests.py
│   └── views.py
├── manage.py
├── poetry.lock
├── project
│   ├── asgi.py
│   ├── __init__.py
│   ├── settings.py
│   ├── urls.py
│   └── wsgi.py
└── pyproject.toml

```
Now poetry works excellent to manage dependencies for whole project.
`poetry build` will build whole project include both applications (with the default version of pyproject.toml).

What if I want additionally build `app1` and `app2` packages into separate builds?
And maybe renamed (because my short app names are not unique on pypi, but I understand that I can achieve this with `name=` and `packages=` in the configuration).

**Reply by sinoroc** [reply]
You want to also be able to build distributions of `app1` and `app2`?

**Comment by abn** [comment]
I believe this will be covered in #2270.
